// s2s-tooltip/src/tooltip.jsx
import React from 'react';
import PropTypes from 'prop-types';
import { fromJS } from "immutable";
import S2SBaseComponent from "s2s-base-class";
import { colorGrid } from 's2s-themes';
import styled from 'styled-components';
import { injectIntl } from 'react-intl';

// Styles the Container
const ToolTipContainer = styled.span`
  background-color : ${colorGrid.gray9};
  border-radius : 18px;
  color : ${colorGrid.gray0};
  font-family: inherit;
  font-size: 14px;
  line-height: 16px;
  padding : 8px 16px;
  position : absolute;
  text-align: center;
  user-select : none;
`;

class ToolTip extends S2SBaseComponent {
  constructor(props) {
    super(props);
    this.displayName = 'ToolTip';

    this.state = {};
  }

  static propTypes = {
    label: PropTypes.string
  }

  static defaultProps = {
    label: "Default Label"
  }

  static filename = 'ToolTip';
  static displayName = 'ToolTip';

  shouldComponentUpdate(nextProps, nextState) {
    return !fromJS(nextProps).equals(fromJS(this.props)) || !fromJS(nextState).equals(fromJS(this.state));
  }


  render() {
    return (
      <ToolTipContainer className={`ToolTipContainer ` + this.props.className}>
        {this.props.label}
      </ToolTipContainer >
    );
  }
}

export default ToolTip;
