// s2s-date-picker/src/viewHeader.js
import React from 'react';
import S2SBaseComponent from 's2s-base-class';
import Button from 's2s-button';
import { fromJS } from 'immutable';
import defaultComponentStyle from "./styles.js";

class ViewHeader extends S2SBaseComponent {
  constructor(props){
      super(props);
      this.displayName = 'ViewHeader';
      this.state = { };

      this.previousCbClick = this.previousCbClick.bind(this);
      this.navTitleCbClick = this.navTitleCbClick.bind(this);
      this.nextCbClick = this.nextCbClick.bind(this);

      this.prefetchStyles();
  }

  shouldComponentUpdate(nextProps, nextState) {
    return !fromJS(nextProps).equals(fromJS(this.props)) || !fromJS(nextState).equals(fromJS(this.state));
  }

  static propTypes = {
    cbHeaderClick : React.PropTypes.func,
    compStyle: React.PropTypes.object,
    headerData : React.PropTypes.string.isRequired
  };

  static defaultProps = {
    cbHeaderClick : ()=>{ console.log('Please provide a value for cbHeaderClick'); },
    headerData : 'Default Header'
  };

  getDefaultStyle(styleName, objectPropState) {
    return defaultComponentStyle(styleName, objectPropState);
  }

  previousCbClick(){
    this.props.cbHeaderClick('previous');
  }

  navTitleCbClick(){
    this.props.cbHeaderClick('title');
  }

  nextCbClick(){
    this.props.cbHeaderClick('next');
  }

  prefetchStyles(){
    this.headerHoverStyle = this.getStyle('headerHoverStyle');
    this.back = this.getStyle('back');
    this.header = this.getStyle('header');
    this.forward = this.getStyle('forward');
    this.viewHeaderDefault = this.getStyle('viewHeaderDefault');
  }

  render(){

    return(
      <div
          style={this.viewHeaderDefault}
          className="navigation-wrapper"
      >
          <Button
              className="prev"
              cbClick={this.previousCbClick}
              ariaLabel="Previous"
              svgType="ChevronLeftIconSVG"
              compStyle={this.back}
          />
          <Button
              className="navigation-title"
              buttonLabel={this.props.headerData}
              cbClick={this.navTitleCbClick}
              tabIndex={0}
              compStyle={this.header}
              key="header"
          />
          <Button
              className="next"
              cbClick={this.nextCbClick}
              ariaLabel="Next"
              svgType="ChevronRightIconSVG"
              compStyle={this.back}
          />
      </div>
    );
  }
}


export default ViewHeader;
