// s2s-select/lib/styles.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var defaultComponentStyle = function defaultComponentStyle(styleName, objectPropState) {
  //console.log(objectPropState.state.itemsVisible);
  var styles = {
    containerStyle: {
      alignItems: 'center',
      backgroundColor: '#f8f9fa',
      borderTop: '1px solid #dee2e6',
      borderLeft: '1px solid #dee2e6',
      borderRight: '1px solid #dee2e6',
      borderBottom: '1px solid #dee2e6',
      borderRadius: '2px',
      display: 'flex',
      fontSize: '14px'

    },
    containerStyleHover: {
      alignItems: 'center',
      backgroundColor: '#e9ecef',
      borderTop: '1px solid #adb5bd',
      borderLeft: '1px solid #adb5bd',
      borderRight: '1px solid #adb5bd',
      borderBottom: '1px solid #adb5bd',
      borderRadius: '2px',
      display: 'flex',
      fontSize: '14px'
    },
    containerStyleFocus: {
      borderTop: '2px solid #228ae6',
      borderLeft: '2px solid #228ae6',
      borderRight: '2px solid #228ae6',
      borderBottom: '2px solid #228ae6'
    },
    containerStyleError: {
      borderTop: '2px solid #e03131',
      borderLeft: '2px solid #e03131',
      borderRight: '2px solid #e03131',
      borderBottom: '2px solid #e03131'
    },
    dropDown: {
      backgroundColor: '#f8f9fa',
      borderTop: '0px',
      borderLeft: '1px solid #dee2e6',
      borderBottom: '1px solid #dee2e6',
      borderRight: '1px solid #dee2e6',
      boxShadow: '0 2px 8px 0 rgba(32, 36, 40, 0.4)',
      cursor: 'pointer',
      //visibility: objectPropState.state.isOpen ? 'visible' : 'hidden',
      fontSize: '14px',
      maxHeight: 36 * objectPropState.props.itemsVisible + 'px', // causes the scrollbar to appear based on the number of items initially seen in the 'drawer'.
      overflowX: 'auto',
      overflowY: 'auto',
      paddingTop: '5px',
      paddingRight: '0px',
      paddingBottom: '5px',
      paddingLeft: '0px',
      position: 'absolute',
      width: 'calc(100% - 2px)',
      top: '38px',
      zIndex: '10'
    },
    dropDownItem: {
      paddingTop: '5px',
      paddingRight: '0px',
      paddingBottom: '5px',
      paddingLeft: '0px',
      lineHeight: '25px',
      marginTop: '0px',
      marginRight: '3px',
      marginBottom: '0px',
      marginLeft: '3px',
      backgroundColor: '#f8f9fa'
    },

    hoverBackgroundColor: {
      backgroundColor: '#e9ecef'
    },

    selectStyle: {
      alignItems: 'center',
      cursor: 'pointer',
      display: 'flex',
      paddingTop: '5px',
      paddingRight: '0px',
      paddingBottom: '5px',
      paddingLeft: '5px',
      // width: '100%'
      justifyContent: 'space-between',
      outline: 'none'
    },

    disabledSelectStyle: {
      alignItems: 'center',
      backgroundColor: ' #f1f3f5',
      color: '#868e96',
      cursor: 'not-allowed',
      display: 'flex',
      paddingTop: '5px',
      paddingRight: '0px',
      paddingBottom: '5px',
      paddingLeft: '5px',
      justifyContent: 'space-between',
      outline: 'none'
    },
    svgContainerStyle: {
      // marginLeft: 'auto',
      // marginRight: '5px'
    },
    svgStyles: {
      height: '24px',
      width: '24px'
    },
    controlStyle: {
      width: '100%'
    },
    menuOptionStyle: {
      width: '100%',
      position: 'relative'
    },
    alertStyles: {
      alignItems: 'center',
      backgroundColor: '#e03131',
      color: '#f8f9fa',
      display: 'flex',
      fontSize: '12px',
      //fontWeight: 600,
      paddingBottom: '10px',
      paddingLeft: '10px',
      paddingRight: '10px',
      paddingTop: '10px'
    },
    alertSVGStyles: {
      width: '21px',
      height: '21px',
      fill: '#f8f9fa',
      paddingRight: '10px'
    }
  };
  return styles[styleName];
};

exports.default = defaultComponentStyle;
