// s2s-select/lib/select.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _templateObject = _taggedTemplateLiteral(["\n  height : 100%;\n  margin-right : 24px;\n  width : 100%;\n"], ["\n  height : 100%;\n  margin-right : 24px;\n  width : 100%;\n"]),
    _templateObject2 = _taggedTemplateLiteral(["\n  color : inherit;\n  display : ", ";\n  flex-direction : column;\n  font-size : 14px;\n  font-weight : 600;\n  line-height : 14px;\n  margin-bottom : 8px;\n  user-select : none;\n  width : 100%;\n"], ["\n  color : inherit;\n  display : ", ";\n  flex-direction : column;\n  font-size : 14px;\n  font-weight : 600;\n  line-height : 14px;\n  margin-bottom : 8px;\n  user-select : none;\n  width : 100%;\n"]),
    _templateObject3 = _taggedTemplateLiteral(["\n    align-items: center;\n    background-color : ", ";\n    border-radius : 4px;\n    box-shadow: 0px 4px 8px rgba(33, 37, 41, 0.4);\n    color : ", ";\n    display : flex;\n    font-size : 14px;\n    line-height : 16px;\n    padding : 8px 12px;\n    z-index : 3;\n\n    // styling necessary for positioning to <Container />\n    // position : absolute;\n    top : ", "; \n    width : calc(100% - 22px ); // must account for padding left/right of <ErrorContainer />  and border of <StyledInput /> or it is not the full width\n"], ["\n    align-items: center;\n    background-color : ", ";\n    border-radius : 4px;\n    box-shadow: 0px 4px 8px rgba(33, 37, 41, 0.4);\n    color : ", ";\n    display : flex;\n    font-size : 14px;\n    line-height : 16px;\n    padding : 8px 12px;\n    z-index : 3;\n\n    // styling necessary for positioning to <Container />\n    // position : absolute;\n    top : ", "; \n    width : calc(100% - 22px ); // must account for padding left/right of <ErrorContainer />  and border of <StyledInput /> or it is not the full width\n"]);

var _react = require("react");

var _react2 = _interopRequireDefault(_react);

var _s2sBaseClass = require("s2s-base-class");

var _s2sBaseClass2 = _interopRequireDefault(_s2sBaseClass);

var _immutable = require("immutable");

var _objectMerge = require("object-merge");

var _objectMerge2 = _interopRequireDefault(_objectMerge);

var _s2sThemes = require("s2s-themes");

var _propTypes = require("prop-types");

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require("styled-components");

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _s2sSvgIcons = require("s2s-svg-icons");

var _reactSelect = require("react-select");

var _reactSelect2 = _interopRequireDefault(_reactSelect);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _taggedTemplateLiteral(strings, raw) { return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var SelectContainer = _styledComponents2.default.div(_templateObject);

var SelectLabel = _styledComponents2.default.div(_templateObject2, function (props) {
  return props.showSelectLabel ? 'flex' : 'none';
});

// Parent div of <ErrorSVGArea /> and <ErrorText />
var ErrorContainer = _styledComponents2.default.div(_templateObject3, _s2sThemes.colorGrid.red7, _s2sThemes.colorGrid.gray0, function (props) {

  var topPosition = '55px'; // has neither hint or a label

  if (props.hasLabel || props.hasHint) {
    topPosition = '69px';
  }

  if (props.hasLabel && props.hasHint) {
    topPosition = '77px';
  }
  return topPosition;
});

var Select = function (_S2SBaseComponent) {
  _inherits(Select, _S2SBaseComponent);

  function Select(props) {
    _classCallCheck(this, Select);

    var _this = _possibleConstructorReturn(this, (Select.__proto__ || Object.getPrototypeOf(Select)).call(this, props));

    _this.state = {
      isSelectOpen: false,
      selectedItem: _this.props.selectedItem ? _this.props.selectedItem : "",
      isFocus: false,
      filteredItems: _this.props.menuItems
    };
    _this.displayName = 'Select';

    //this.renderMenuItems = this.renderMenuItems.bind(this);
    _this.selectClicked = _this.selectClicked.bind(_this);
    _this.menuItemClicked = _this.menuItemClicked.bind(_this);
    //this.menuItemKeyDown = this.menuItemKeyDown.bind(this);
    _this.selectKeyPress = _this.selectKeyPress.bind(_this);
    //this.getCaretSVG = this.getCaretSVG.bind(this);
    //this.selectedItemClick = this.selectedItemClick.bind(this);
    //this.selectedItemKeyPress = this.selectedItemKeyPress.bind(this);
    _this.handleFocus = _this.handleFocus.bind(_this);
    _this.handleBlur = _this.handleBlur.bind(_this);

    return _this;
  }

  _createClass(Select, [{
    key: "shouldComponentUpdate",
    value: function shouldComponentUpdate(nextProps, nextState) {
      return !(0, _immutable.fromJS)(nextProps).equals((0, _immutable.fromJS)(this.props)) || !(0, _immutable.fromJS)(nextState).equals((0, _immutable.fromJS)(this.state));
    }
  }, {
    key: "getSelectedItemIndex",
    value: function getSelectedItemIndex(menuitems) {
      var _this2 = this;

      var selectedIndex = menuitems.reduce(function (previous, current, index) {
        if (!previous) {

          if (current.value === _this2.state.selectedItem) {
            return index;
          }
        }
        return previous;
      }, undefined);

      return selectedIndex;
    }
  }, {
    key: "selectClicked",
    value: function selectClicked() {
      var _this3 = this;

      if (this.props.isDisabled === false) {
        this.setState(function (prevState) {
          return _extends({}, prevState, { isSelectOpen: !_this3.state.isSelectOpen });
        });
      }
    }
  }, {
    key: "selectKeyPress",
    value: function selectKeyPress(e) {
      var _this4 = this;

      //console.log('kkkkk', e, e.keyCode, e.key,  e.charCode);
      if (e.charCode > 0 || e.charCode < 123) {
        var myKeyText = e.key;
        // console.log('find first one ', e.key );
        var bFound = false;
        var fItem = void 0;
        var sItem = this.props.menuItems.reduce(function (p, c) {
          if (!p) {
            //console.log(c.label.toLowerCase()[0], e.key.toLowerCase() );
            if (c.label.toLowerCase().substring(0, myKeyText.length) === myKeyText.toLowerCase()) {
              if (!fItem) {
                fItem = c.value;
              }
              if (c.value !== _this4.state.selectedItem) {
                //console.log('SSSSS', this.state.selectedItem, bFound, c.label, c.value);
                if (!_this4.state.selectedItem || _this4.state.selectedItem && bFound) {
                  return c.value;
                }
              } else {
                bFound = true;
              }
            }
          }
          return p;
        }, undefined);
        var filteredItems = this.props.menuItems.filter(function (item) {
          return item.label.toLowerCase().substring(0, myKeyText.length) === myKeyText.toLowerCase();
        });

        //console.log('sssss', sItem, fItem, this.state.selectedItem,  myKeyText, filteredItems, filteredItems.length);
        if (sItem) {
          this.setState(function (prevState) {
            return _extends({}, prevState, { selectedItem: sItem, "filteredItems": filteredItems, isSelectOpen: true });
          });
        } else if (fItem) {
          this.setState(function (prevState) {
            return _extends({}, prevState, { selectedItem: fItem, "filteredItems": filteredItems, isSelectOpen: true });
          });
        } else {
          this.setState(function (prevState) {
            return _extends({}, prevState, { selectedItem: fItem, "filteredItems": _this4.props.menuItems, isSelectOpen: true });
          });
        }
      }
      if (e.charCode == 13) {
        this.setState(function (prevState) {
          return _extends({}, prevState, { "filteredItems": _this4.props.menuItems, isSelectOpen: !_this4.state.isSelectOpen });
        });
      }
    }
  }, {
    key: "menuItemClicked",
    value: function menuItemClicked(item) {
      var _this5 = this;

      this.setState(function (prevState) {
        return _extends({}, prevState, { selectedItem: item.value, "filteredItems": _this5.props.menuItems, isSelectOpen: false });
      });
      //console.log('xxxxx', item, this.props );
      // adding doValidate if we have errorText so they clear errorText
      if (this.props.errorText && this.props.errorText.length > 0) {
        item.doValidate = true;
      } else {
        item.doValidate = false;
      }
      this.props.cbOnValueChange(item);
    }

    // menuItemKeyDown(e, item) {
    //   if (e.charCode == 13) {
    //     this.setState((prevState) => {
    //       return { ...prevState, selectedItem: item.value, "filteredItems": this.props.menuItems, isSelectOpen: false };
    //     });
    //     this.props.cbOnValueChange(item);
    //   }
    // }

    // selectedItemClick() {
    //   this.setState((prevState) => {
    //     return { ...prevState, keyText: "", "filteredItems": this.props.menuItems, isSelectOpen: false };
    //   });
    // }

    // selectedItemKeyPress(e) {
    //   if (e.charCode == 13) {
    //     this.setState((prevState) => {
    //       return { ...prevState, isSelectOpen: false };
    //     });
    //   }
    // }

  }, {
    key: "handleFocus",
    value: function handleFocus() {
      if (!this.props.isDisabled) {
        this.setState(function (prevState) {
          return (0, _objectMerge2.default)({}, prevState, {
            'isFocus': true
          });
        });
      }
    }
  }, {
    key: "handleBlur",
    value: function handleBlur() {
      if (!this.props.isDisabled) {
        this.setState(function (prevState) {
          return (0, _objectMerge2.default)({}, prevState, {
            'isFocus': false
          });
        });
        this.props.cbOnBlur(this.state.selectedItem);
      }
    }

    // renderMenuItems(menuItems, selectedItemIndex) {
    //   //console.log('render menu items', menuItems);
    //   return menuItems.map((item, index) => {

    //     let menuItem;
    //     if (index === selectedItemIndex) {
    //       menuItem = <SelectedMenuItem tabIndex={0} role="button" key={item.label + index} onClick={this.selectedItemClick} onKeyPress={this.selectedItemKeyPress}>{item.label}</SelectedMenuItem>;
    //     } else {
    //       menuItem = <SelectMenuItem key={item.label + index} text={item.label} item={item} onItemClick={this.menuItemClicked} onItemKeyPress={this.menuItemKeyDown} />;
    //     }
    //     return menuItem;
    //   });
    // }

    // Render the error area

  }, {
    key: "renderErrorArea",
    value: function renderErrorArea() {

      if (this.props.hasError === true && this.props.errorText !== undefined && this.state.isFocus === false) {
        return _react2.default.createElement(
          ErrorContainer,
          {
            className: "ErrorContainer",
            hasLabel: this.props.inputLabel !== undefined,
            hasHint: this.props.inputHint !== undefined
          },
          _react2.default.createElement(_s2sSvgIcons.AlertIconSVG, {
            className: "AlertIconSVG",
            svgStyle: {
              fill: _s2sThemes.colorGrid.gray0,
              height: '24px',
              width: '24px',
              paddingRight: '12px'
            }
          }),
          _react2.default.createElement(
            "span",
            { className: "ErrorText" },
            this.props.errorText
          )
        );
      }
    }

    // //If you know of a better way to do this please let me know! CS
    // getCaretSVG() {
    //   let caretSVG;
    //   let caretUp = <CaretUpIconSVG svgStyle={{ height: '16px', marginRight: '12px', width: '16px', fill: colorGrid.gray9 }} />;
    //   let caretDown = <CaretDownIconSVG svgStyle={{ height: '16px', marginRight: '12px', width: '16px', fill: colorGrid.gray9 }} />;

    //   if (this.state.isSelectOpen) {
    //     if (this.props.openAbove) {
    //       caretSVG = caretDown;
    //     } else {
    //       caretSVG = caretUp;
    //     }
    //   } else {
    //     if (this.props.openAbove) {
    //       caretSVG = caretUp;
    //     } else {
    //       caretSVG = caretDown;
    //     }
    //   }
    //   return caretSVG;
    // }

  }, {
    key: "render",
    value: function render() {
      var _this6 = this;

      //onst dynamicSVG = this.getCaretSVG();
      //const selectedItemIndex = this.getSelectedItemIndex(this.props.menuItems);

      //console.log('select props', this.props.errorText);

      //const selectMenuHeight = Math.min(this.props.menuItems.length * MENU_ITEM_HEIGHT, this.props.itemsVisible * MENU_ITEM_HEIGHT);
      return _react2.default.createElement(
        SelectContainer,
        { className: this.props.className },
        _react2.default.createElement(
          SelectLabel,
          { showSelectLabel: this.props.showSelectLabel },
          this.props.selectLabel
        ),
        _react2.default.createElement(_reactSelect2.default, { options: this.props.menuItems,
          onBlur: this.handleBlur,
          onFocus: this.handleFocus,
          onChange: this.menuItemClicked, isDisabled: this.props.isDisabled,
          value: this.props.menuItems.filter(function (i) {
            return i.value === _this6.props.selectedItem;
          }),
          placeholder: this.props.placeholder,
          styles: {
            "menu": function menu(style) {
              return _extends({}, style, { "z-index": 99 });
            }
          } }),
        this.renderErrorArea()
      );
    }
  }]);

  return Select;
}(_s2sBaseClass2.default);

Select.propTypes = {
  //openAbove: PropTypes.bool,
  isDisabled: _propTypes2.default.bool,
  menuItems: _propTypes2.default.array,
  cbOnValueChange: _propTypes2.default.func,
  cbOnBlur: _propTypes2.default.func,
  selectLabel: _propTypes2.default.string,
  selectedItem: _propTypes2.default.any,
  //itemsVisible: PropTypes.number,
  placeholder: _propTypes2.default.string,
  showSelectLabel: _propTypes2.default.bool
};
Select.defaultProps = {
  //openAbove: false,
  isDisabled: false,
  cbOnValueChange: function cbOnValueChange() {
    console.log("cbOnValueChange is not defined");
  },
  cbOnBlur: function cbOnBlur() {
    console.log('cbOnBlur is not defined');
  },
  selectedItem: "",
  selectLabel: undefined,
  //itemsVisible: 3,
  placeholder: "Please select a value",
  showSelectLabel: true
};
exports.default = Select;
