// s2s-date-picker/lib/styles.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var defaultComponentStyle = function defaultComponentStyle(styleName, objectPropState) {
  var styles = {
    //// INDEX
    // Component Wrapper
    datePicker: {
      alignItems: 'center',
      backgroundColor: '#f8f9fa',
      color: '#212529',
      cursor: 'pointer',
      display: 'flex',
      flexDirection: 'column'
      // width: '240px' //
    },
    // area that contains Positioned Buttons, Calendar Button, and Input.
    topContainer: {
      display: 'flex',
      position: 'relative',
      width: '100%' // flex 1 collapses s2s-input
    },
    positionedButtonsWrapper: {
      alignItems: 'center',
      backgroundColor: '#f8f9fa',
      borderRadius: '2px',
      cursor: 'pointer',
      //display:  objectPropState.state.dropDownDisplayed ? 'flex' : 'none', //
      flex: 1,
      //paddingRight: objectPropState.props.displayTime? '0px':'100px', //
      position: 'absolute',
      //top: objectPropState.state.dropDownDisplayed == true ? '7px' :'-1000px', //
      left: '50px'
    },

    // Button that is hidden off screen until in picking view
    dateButton: {
      backgroundColor: '#f8f9fa',
      border: '1px solid transparent',
      borderRadius: '2px',
      flex: '1',
      fontSize: '12px',
      height: '30px',
      marginRight: '5px',
      paddingBottom: '0px',
      paddingLeft: '0px',
      paddingRight: '0px',
      paddingTop: '0px',
      width: '75px'
    },
    dateButtonActive: {
      backgroundColor: '#228ae6',
      border: '1px solid #228ae6',
      borderRadius: '2px',
      color: '#f8f9fa',
      flex: '1',
      fontSize: '12px',
      height: '30px',
      marginRight: '5px',
      paddingBottom: '0px',
      paddingLeft: '0px',
      paddingRight: '0px',
      paddingTop: '0px',
      width: '75px'
    },
    dateButtonHover: {
      backgroundColor: '#e9ecef',
      border: '1px solid #e9ecef',
      borderRadius: '2px',
      color: '#212529',
      flex: '1',
      fontSize: '12px',
      height: '30px',
      marginRight: '5px',
      paddingBottom: '0px',
      paddingLeft: '0px',
      paddingRight: '0px',
      paddingTop: '0px',
      width: '75px'
    },

    // Button that is hidden off screen until in picking view
    timeButton: {
      backgroundColor: '#f8f9fa',
      border: '1px solid transparent',
      borderRadius: '2px',
      flex: '1',
      fontSize: '12px',
      height: '30px',
      marginRight: '0px',
      paddingBottom: '0px',
      paddingLeft: '0px',
      paddingRight: '0px',
      paddingTop: '0px',
      width: '75px'
    },
    timeButtonActive: {
      backgroundColor: '#228ae6',
      border: '1px solid #228ae6',
      borderRadius: '2px',
      color: '#f8f9fa',
      flex: '1',
      fontSize: '12px',
      height: '30px',
      marginRight: '0px',
      paddingBottom: '0px',
      paddingLeft: '0px',
      paddingRight: '0px',
      paddingTop: '0px',
      width: '75px'
    },
    timeButtonHover: {
      backgroundColor: '#e9ecef',
      border: '1px solid #e9ecef',
      borderRadius: '2px',
      color: '#212529',
      flex: '1',
      fontSize: '12px',
      height: '30px',
      marginRight: '0px',
      paddingBottom: '0px',
      paddingLeft: '0px',
      paddingRight: '0px',
      paddingTop: '0px',
      width: '75px'
    },
    // input area that is initially displayed.
    inputStyle: {
      alertSvg: {},
      alertSvgArea: {},
      disabledInput: {},
      errorText: {},
      inputAlert: {},
      inputField: {},
      inputFieldNoSVG: {
        paddingBottom: '5px',
        paddingTop: '5px',
        width: '100%'
      },
      inputFieldWithError: {},
      inputFieldWithErrorNoSVG: {},
      inputFieldWithHover: {
        borderBottomRightRadius: '0px'
      },
      inputFieldWithHoverNoSVG: {
        //flex: 1,
        paddingBottom: '5px',
        paddingTop: '5px',
        width: '100%'
      },
      inputComponent: {},
      inputContainer: {
        border: '1px solid transparent'
      },
      inputContainerWithFocus: {
        border: '1px solid #228ae6'
      },
      svgArea: {
        // backgroundColor: '#f1f3f5'
        // border: objectPropState.state.dropDownDisplayed ? '1px solid #228ae6' : '1px solid #dee2e6'
      },
      noSVGArea: {
        //fill: objectPropState.state.dropDownDisplayed ? "#f8f9fa": "#212529"
      },
      svgAreaWithError: {},
      svgAreaWithFocus: {
        // backgroundColor: '#228ae6',
        border: '1px solid #228ae6'
      },
      svgAreaWithHover: {},
      svgStyle: {
        fill: '#212529'
      },
      svgStyleChange: {
        fill: '#f8f9fa',
        height: '24px',
        width: '24px'
      },
      svgStyleWithHover: {
        fill: '#212529'
      },
      noSvgStyle: {}
    },
    dropdownWrapper: {
      backgroundColor: '#f8f9fa',
      width: '100%'
    },

    ////CALENDER

    calendarWrapper: {
      userSelect: 'none'
    },
    todayWrapper: {},
    today: {
      alignItems: 'center',
      display: 'flex',
      justifyContent: 'center',
      padding: '5px'
    },
    todayButton: {
      backgroundColor: '#4ab7e2',
      borderBottom: 'none',
      borderLeft: 'none',
      borderRight: 'none',
      borderTop: 'none',
      color: '#f8f9fa'
    },

    ////CELL

    cellWrapper: {
      alignItems: 'center',
      borderBottom: '1px solid transparent',
      borderLeft: '1px solid transparent',
      borderRight: '1px solid transparent',
      borderTop: '1px solid transparent',
      borderRadius: '2px',
      display: 'flex',
      flex: '1 1 13%',
      fontSize: '16px',
      height: '30px',
      justifyContent: 'center'
      // paddingBottom: '5px',
      // paddingTop: '5px'
    },
    isSelected: {
      backgroundColor: '#228ae6',
      color: '#f8f9fa'
    },
    cellHoverStyle: {
      // backgroundColor: '#fbfbfb',
      // borderBottom: '1px solid #efefef',
      // borderLeft: '1px solid #efefef',
      // borderRight: '1px solid #efefef',
      // borderTop: '1px solid #efefef',
      // color: '#228ae6'
    },

    ////DAYVIEW
    daysTitleWrapper: {
      alignItems: 'center',
      display: 'flex',
      flex: 1,
      marginBottom: '5px',
      paddingBottom: '10px'
      //paddingTop: '10px'
    },
    daysTitle: {
      cellWrapper: {},
      isSelected: {},
      hoverStyle: {},
      alignItems: 'center',
      display: 'flex',
      flex: 1,
      fontWeight: '500',
      justifyContent: 'center'
    },
    viewHeaderDayView: {
      viewHeader: {},
      back: {},
      header: {},
      next: {},
      hoverStyle: {}
    },
    daysWrapper: {
      display: 'flex',
      flex: '1',
      flexWrap: 'wrap',
      justifyContent: 'space-between',
      paddingBottom: '10px',
      paddingLeft: '20px',
      paddingRight: '20px',
      paddingTop: '10px'
    },
    // days that do not belong to month being displayed
    lonerDays: {
      cellWrapper: {},
      isSelected: {},
      hoverStyle: {},
      color: '#dee2e6'
    },
    selectedDay: {
      // backgroundColor: '#4ab7e2',
      // color: '#f8f9fa'
    },
    days: {},

    ////MONTHVIEW

    monthWrapper: {
      display: 'flex',
      flex: '1',
      flexWrap: 'wrap',
      justifyContent: 'space-between',
      paddingBottom: '10px',
      paddingLeft: '20px',
      paddingRight: '20px',
      paddingTop: '10px'
    },
    month: {
      cellWrapper: {},
      isSelected: {},
      hoverStyle: {},
      backgroundColor: '#f1f3f5',
      color: '#212529',
      flex: '1 1 18%',
      marginBottom: '10px',
      marginLeft: '5px',
      marginRight: '5px'

    },
    viewHeaderMonthView: {
      viewHeader: {
        viewHeader: {
          display: 'flex'
        },
        back: {},
        header: {},
        next: {},
        hoverStyle: {}
      }
    },

    ////timeView

    timeWrapper: {
      display: 'flex',
      flex: '1',
      flexDirection: 'column',
      flexWrap: 'wrap',
      justifyContent: 'center'
    },
    timeFormatWrapper: {
      backgroundColor: '#f1f3f5',
      boxShadow: '0 1px 0 0 #ced4da',
      display: 'flex',
      justifyContent: 'center',
      marginBottom: '1px',
      paddingBottom: '5px',
      paddingTop: '5px'

    },
    topContainerTimeView: {
      alignItems: 'center',
      display: 'flex'
    },
    hourWrapper: {
      marginLeft: '10px',
      marginRight: '10px'
    },
    hour: {
      height: '32px',
      paddingLeft: '5px',
      width: '32px'
    },
    minuteWrapper: {
      marginLeft: '10px',
      marginRight: '10px'
    },
    minute: {
      height: '32px',
      paddingLeft: '5px',
      width: '32px'
    },
    dateTime: {
      buttonStyles: {
        backgroundColor: '#f8f9fa',
        borderBottom: 'solid 1px #dee2e6',
        borderLeft: 'solid 1px #dee2e6',
        borderRight: 'solid 1px #dee2e6',
        borderTop: 'solid 1px #dee2e6',
        borderRadius: 'none',
        fontSize: '12px',
        outline: 'none',
        height: '32px',
        paddingBottom: '0px',
        paddingLeft: '0px',
        paddingRight: '0px',
        paddingTop: '0px',
        width: '32px'
      },
      svgStyles: {},
      labelStyles: {
        color: '#212529'
      }
    },
    buttonWrapperStyles: {
      paddingBottom: '10px',
      paddingLeft: '20px',
      paddingRight: '20px',
      paddingTop: '10px'
    },
    buttonArea: {
      //alignItems: 'center',
      //alignSelf: 'center',
      display: 'flex',
      flexWrap: 'wrap'
      //justifyContent: 'center',
      //width: '75%'
    },
    button: {
      //width: '30px'
      flex: '1 1 25%',
      backgroundColor: '#f1f3f5',
      color: '#212529',
      marginBottom: '10px',
      marginLeft: '5px',
      marginRight: '5px'
    },
    backButton: {
      buttonStyles: {
        backgroundColor: '#f1f3f5',
        borderBottom: 'none',
        borderLeft: 'none',
        borderRight: 'none',
        borderTop: 'none',
        borderRadius: '2px',
        marginBottom: '10px',
        marginLeft: '5px',
        marginRight: '5px',
        paddingBottom: '0px',
        paddingLeft: '0px',
        paddingRight: '0px',
        paddingTop: '0px',
        flex: '1 0 25%',
        width: '60px'
      },
      labelStyles: {},
      svgStyles: {
        height: '18px',
        width: '24px'
      }
      // hoverButtonStyles: {
      //   backgroundColor: '#e9ecef'
      // }
    },

    ////VIEWHEADER
    viewHeaderDefault: {
      alignItems: 'center',
      backgroundColor: '#f1f3f5',
      boxShadow: '0 1px 0 0 #ced4da',
      cursor: 'pointer',
      display: 'flex',
      fontSize: '16px',
      justifyContent: 'space-around',
      paddingBottom: '5px',
      paddingTop: '5px',
      userSelect: 'none'
    },
    back: {
      buttonStyles: {
        backgroundColor: 'transparent',
        borderBottom: '1px solid transparent',
        borderLeft: '1px solid transparent',
        borderRight: '1px solid transparent',
        borderTop: '1px solid transparent',
        paddingBottom: '0px',
        paddingLeft: '0px',
        paddingRight: '0px',
        paddingTop: '0px'
      },
      labelStyles: {},
      svgStyles: {
        fill: '#212529'
      },
      hoverSvgStyles: {
        fill: '#4ab7e2'
      }
    },
    header: {
      buttonStyles: {
        backgroundColor: 'transparent',
        borderBottom: '1px solid transparent',
        borderLeft: '1px solid transparent',
        borderRight: '1px solid transparent',
        borderTop: '1px solid transparent',
        paddingBottom: '0px',
        paddingTop: '0px'
      },
      svgStyles: {},
      labelStyles: {},
      hoverButtonStyles: {
        //backgroundColor: '#228ae6'
      },
      hoverLabelStyles: {
        color: '#4ab7e2'
      },
      hoverSvgStyles: {},
      focusButtonStyles: {
        //backgroundColor: '#228ae6'
      },
      focusLabelStyles: {
        //color: '#f8f9fa'
      },
      focusSvgStyles: {}

    },
    forward: {
      buttonStyles: {
        backgroundColor: 'transparent',
        borderBottom: '1px solid transparent',
        borderLeft: '1px solid transparent',
        borderRight: '1px solid transparent',
        borderTop: '1px solid transparent',
        paddingBottom: '0px',
        paddingLeft: '0px',
        paddingRight: '0px',
        paddingTop: '0px'
      },
      labelStyles: {},
      svgStyles: {
        fill: '#212529'
      },
      hoverSvgStyles: {
        fill: '#4ab7e2'
      }
    },

    ////YEARVIEW
    yearWrapper: {
      display: 'flex',
      flex: '1',
      flexWrap: 'wrap',
      justifyContent: 'space-between',
      paddingBottom: '10px',
      paddingLeft: '20px',
      paddingRight: '20px',
      paddingTop: '10px'
    },
    viewHeaderYearView: {
      viewHeader: {},
      back: {},
      header: {},
      forward: {},
      hoverStyle: {}
    },
    year: {
      backgroundColor: '#f1f3f5',
      color: '#212529',
      flex: '1 1 18%',
      marginBottom: '10px',
      marginLeft: '5px',
      marginRight: '5px'
    }
  };
  return styles[styleName];
};

exports.default = defaultComponentStyle;
