// s2s-date-picker/lib/viewHeader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _s2sBaseClass = require('s2s-base-class');

var _s2sBaseClass2 = _interopRequireDefault(_s2sBaseClass);

var _s2sButton = require('s2s-button');

var _s2sButton2 = _interopRequireDefault(_s2sButton);

var _immutable = require('immutable');

var _styles = require('./styles.js');

var _styles2 = _interopRequireDefault(_styles);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ViewHeader = function (_S2SBaseComponent) {
  _inherits(ViewHeader, _S2SBaseComponent);

  function ViewHeader(props) {
    _classCallCheck(this, ViewHeader);

    var _this = _possibleConstructorReturn(this, (ViewHeader.__proto__ || Object.getPrototypeOf(ViewHeader)).call(this, props));

    _this.displayName = 'ViewHeader';
    _this.state = {};

    _this.previousCbClick = _this.previousCbClick.bind(_this);
    _this.navTitleCbClick = _this.navTitleCbClick.bind(_this);
    _this.nextCbClick = _this.nextCbClick.bind(_this);

    _this.prefetchStyles();
    return _this;
  }

  _createClass(ViewHeader, [{
    key: 'shouldComponentUpdate',
    value: function shouldComponentUpdate(nextProps, nextState) {
      return !(0, _immutable.fromJS)(nextProps).equals((0, _immutable.fromJS)(this.props)) || !(0, _immutable.fromJS)(nextState).equals((0, _immutable.fromJS)(this.state));
    }
  }, {
    key: 'getDefaultStyle',
    value: function getDefaultStyle(styleName, objectPropState) {
      return (0, _styles2.default)(styleName, objectPropState);
    }
  }, {
    key: 'previousCbClick',
    value: function previousCbClick() {
      this.props.cbHeaderClick('previous');
    }
  }, {
    key: 'navTitleCbClick',
    value: function navTitleCbClick() {
      this.props.cbHeaderClick('title');
    }
  }, {
    key: 'nextCbClick',
    value: function nextCbClick() {
      this.props.cbHeaderClick('next');
    }
  }, {
    key: 'prefetchStyles',
    value: function prefetchStyles() {
      this.headerHoverStyle = this.getStyle('headerHoverStyle');
      this.back = this.getStyle('back');
      this.header = this.getStyle('header');
      this.forward = this.getStyle('forward');
      this.viewHeaderDefault = this.getStyle('viewHeaderDefault');
    }
  }, {
    key: 'render',
    value: function render() {

      return _react2.default.createElement(
        'div',
        {
          style: this.viewHeaderDefault,
          className: 'navigation-wrapper'
        },
        _react2.default.createElement(_s2sButton2.default, {
          className: 'prev',
          cbClick: this.previousCbClick,
          ariaLabel: 'Previous',
          svgType: 'ChevronLeftIconSVG',
          compStyle: this.back
        }),
        _react2.default.createElement(_s2sButton2.default, {
          className: 'navigation-title',
          buttonLabel: this.props.headerData,
          cbClick: this.navTitleCbClick,
          tabIndex: 0,
          compStyle: this.header,
          key: 'header'
        }),
        _react2.default.createElement(_s2sButton2.default, {
          className: 'next',
          cbClick: this.nextCbClick,
          ariaLabel: 'Next',
          svgType: 'ChevronRightIconSVG',
          compStyle: this.back
        })
      );
    }
  }]);

  return ViewHeader;
}(_s2sBaseClass2.default);

ViewHeader.propTypes = {
  cbHeaderClick: _react2.default.PropTypes.func,
  compStyle: _react2.default.PropTypes.object,
  headerData: _react2.default.PropTypes.string.isRequired
};
ViewHeader.defaultProps = {
  cbHeaderClick: function cbHeaderClick() {
    console.log('Please provide a value for cbHeaderClick');
  },
  headerData: 'Default Header'
};
exports.default = ViewHeader;
