// s2s-date-picker/lib/yearView.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _s2sBaseClass = require('s2s-base-class');

var _s2sBaseClass2 = _interopRequireDefault(_s2sBaseClass);

var _immutable = require('immutable');

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

require('moment-range');

var _cell = require('./cell');

var _cell2 = _interopRequireDefault(_cell);

var _viewHeader = require('./viewHeader');

var _viewHeader2 = _interopRequireDefault(_viewHeader);

var _styles = require('./styles.js');

var _styles2 = _interopRequireDefault(_styles);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var YearView = function (_S2SBaseComponent) {
  _inherits(YearView, _S2SBaseComponent);

  function YearView(props) {
    _classCallCheck(this, YearView);

    var _this = _possibleConstructorReturn(this, (YearView.__proto__ || Object.getPrototypeOf(YearView)).call(this, props));

    _this.displayName = 'YearView';
    _this.state = { dateVal: _this.props.dateVal };

    _this.handleTitleClick = _this.handleTitleClick.bind(_this);
    _this.cellClick = _this.cellClick.bind(_this);
    _this.prev = _this.prev.bind(_this);
    _this.next = _this.next.bind(_this);
    _this.getYears = _this.getYears.bind(_this);
    _this.checkIfYearDisabled = _this.checkIfYearDisabled.bind(_this);
    _this.yearsOnKeyPress = _this.yearsOnKeyPress.bind(_this);
    _this.moment = (0, _moment2.default)();

    _this.prefetchStyles();
    return _this;
  }

  _createClass(YearView, [{
    key: 'componentWillMount',
    value: function componentWillMount() {
      this.getYears();
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (nextProps.hasOwnProperty('dateVal')) {
        this.setState(_extends({}, this.state, { dateVal: nextProps.dateVal }));
      }
    }
  }, {
    key: 'shouldComponentUpdate',
    value: function shouldComponentUpdate(nextProps, nextState) {
      return !(0, _immutable.fromJS)(nextProps).equals((0, _immutable.fromJS)(this.props)) || !(0, _immutable.fromJS)(nextState).equals((0, _immutable.fromJS)(this.state));
    }
  }, {
    key: 'getDefaultStyle',
    value: function getDefaultStyle(styleName, objectPropState) {
      return (0, _styles2.default)(styleName, objectPropState);
    }
  }, {
    key: 'cellClick',
    value: function cellClick(e) {
      var year = parseInt(e.target.innerHTML, 10);
      var date = this.state.dateVal.clone().year(year);
      if (this.checkIfYearDisabled(date)) return;
      this.props.cbPrevView(date);
    }
  }, {
    key: 'checkIfYearDisabled',
    value: function checkIfYearDisabled(year) {
      return year.clone().endOf('year').isBefore(this.props.minDate, 'day') || year.clone().startOf('year').isAfter(this.props.maxDate, 'day');
    }
  }, {
    key: 'getYears',
    value: function getYears() {
      var _this2 = this;

      var now = this.state.dateVal; // TODO: fix bad variable name
      var start = now.clone().subtract(5, 'year');
      var end = now.clone().add(6, 'year');
      var currYear = now.year();
      var items = [];

      this.moment.range(start, end).by('years', function (year) {

        items.push({
          label: year.format('YYYY'),
          disabled: _this2.checkIfYearDisabled(year),
          curr: currYear === year.year()
        });
      });

      return items;
    }
  }, {
    key: 'next',
    value: function next() {
      var nextDate = this.state.dateVal.clone().add(10, 'years');
      if (this.props.maxDate && nextDate.isAfter(this.props.maxDate, 'day')) {
        nextDate = this.props.maxDate;
      }
      this.props.cbSetDate(nextDate);
    }
  }, {
    key: 'prev',
    value: function prev() {
      var prevDate = this.state.dateVal.clone().subtract(10, 'years');
      if (this.props.minDate && prevDate.isBefore(this.props.minDate, 'day')) {

        prevDate = this.props.minDate;
      }
      this.props.cbSetDate(prevDate);
    }
  }, {
    key: 'handleTitleClick',
    value: function handleTitleClick(type) {
      switch (type) {
        case 'previous':
          this.prev();
          break;
        case 'next':
          this.next();
          break;
      }
    }
  }, {
    key: 'yearsOnKeyPress',
    value: function yearsOnKeyPress(e) {
      if (e.charCode == '13') {
        this.cellClick(e);
      }
    }
  }, {
    key: 'prefetchStyles',
    value: function prefetchStyles() {
      this.yearStyles = this.getStyle('year');
      this.viewHeaderYearView = this.getStyle('viewHeaderYearView');
      this.yearWrapper = this.getStyle('yearWrapper');
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      var years = this.getYears();
      var currentDate = [years[0].label, years[years.length - 1].label].join('-');

      var yearsCells = years.map(function (item, i) {
        return _react2.default.createElement(_cell2.default, {
          dateVal: item.label,
          key: i,
          compStyle: { cellWrapper: _extends({}, _this3.yearStyles) }
        });
      });

      return _react2.default.createElement(
        'div',
        { className: 'years-view' },
        _react2.default.createElement(_viewHeader2.default, {
          compStyle: this.viewHeaderYearView,
          cbHeaderClick: this.handleTitleClick,
          headerData: currentDate
        }),
        _react2.default.createElement(
          'div',
          {
            className: 'years',
            onClick: this.cellClick,
            style: this.yearWrapper,
            onKeyPress: this.yearsOnKeyPress
          },
          yearsCells
        )
      );
    }
  }]);

  return YearView;
}(_s2sBaseClass2.default);

YearView.propTypes = {
  cbPrevView: _react2.default.PropTypes.func,
  cbSetDate: _react2.default.PropTypes.func,
  compStyle: _react2.default.PropTypes.object,
  dateVal: _react2.default.PropTypes.object.isRequired,
  maxDate: _react2.default.PropTypes.object, // moment date object
  minDate: _react2.default.PropTypes.object // moment date object
};
YearView.defaultProps = {
  cbPrevView: function cbPrevView() {
    console.log('Please provide a callback function for cbPrevView');
  },
  cbSetDate: function cbSetDate() {
    console.log('Please provide a callback function for cbSetDate');
  },
  dateVal: {},
  minDate: (0, _moment2.default)().add(-10000, 'years'),
  maxDate: (0, _moment2.default)().add(10000, 'years')
};
exports.default = YearView;
